bot.go

```diff
@@ -0,0 +1,306 @@
+
+		// NOTE that the following errors will cause any commands still in the queue to error
+		// out while it reconnects.
+		if err != nil {
+			b.setReconnect()
+
+			if err == io.EOF {
+				b.Log(logger.SeverityError, "Reached EOF. Reconnecting")
+			} else if err != nil {
+				b.Log(logger.SeverityError, "Error reading. "+err.Error())
+			}
+			break
+		}
+
+		// Add to the buffer.
+		buf = append(buf, tmp[:n]...)
+
+		// Find \r\n
+		idx := bytes.Index(buf, []byte("\r\n"))
+		curr := strings.TrimSpace(string(buf[:idx+2]))
+		buf = buf[idx+2:]
+
+		switch {
+		case strings.Index(curr, "PING") > -1:
+			b.setSendPong()
+		}
+
+		log.Print("MSG: ", curr)
+	}
+}
+
+// Connect is used to connect and authenticate to IIRC.
+// It requires the oauth access token as the password in the format oath:<token>.
+// In the case of failure, uses exponential backoff.
+//
+// https://dev.twitch.tv/docs/irc/guide
+func (b *IIRCBot) Connect(username, password string) error {
+	b.username = username
+	b.password = password
+
+	delay := time.Second / 2
+
+	for {
+		if delay > time.Second*16 {
+			b.Log(logger.SeverityError, "Timed out.")
+			return ErrTimeout
+		}
+
+		err := b.connect(username, password)
+		if err == nil {
+			return nil
+		}
+
+		b.Log(logger.SeverityError, fmt.Sprintf("Could not login. Waiting for %d seconds. %s", delay*2, err.Error()))
+		delay = delay * 2
+	}
+}
+
+// Reconnect is used to refresh the password used by the IIRCBot. Since it is using OAUTH,
+// the password will expire and will need to be replaced. In this case, all commands will need to
+// wait until reconnection is finished.
+func (b *IIRCBot) Reconnect(username, password string) {
+	b.Queue <- &IIRCBotConnect{
+		bot:      b,
+		Username: username,
+		Password: password,
+	}
+}
+
+// Send sends a command through the IIRCBot. It does this by scheduling on the queue.
+func (b *IIRCBot) Send(cmd, args string) {
+	b.Queue <- &IIRCBotMessage{
+		bot:  b,
+		Cmd:  cmd,
+		Args: args,
+	}
+}
+
+func (b *IIRCBot) Say(channel, msg string) {
+	b.Queue <- &IIRCBotSay{
+		bot:     b,
+		Channel: channel,
+		Message: msg,
+	}
+}
+
+// Run runs the service which actually posts the messages on the channel.
+func (b *IIRCBot) Run(die <-chan struct{}) {
+	delay := b.MsgDelay
+	if delay == 0 {
+		delay = DefaultRate
+	}
+
+	// Allow a queue of 50 items.
+	b.Queue = make(chan IIRCBotCommand, 50)
+
+	go func() {
+		for {
+			select {
+			case <-die:
+				close(b.Queue)
+				return
+
+			case <-time.After(delay):
+			}
+
+			// Check flags first! If we need to reconnect, we should do that here.
+			if b.NeedsReconnect() {
+				b.Reconnect(b.username, b.password)
+				continue
+			}
+			if b.needsSendPong() {
+				if err := b.send(CmdPong, ":tmi.twitch.tv"); err != nil {
+					b.Log(logger.SeverityError, "Could not send pong. "+err.Error())
+				}
+			}
+
+			// Run a queue command if there are any. 
+			if len(b.Queue) > 0 {
+				cmd := <-b.Queue
+				if err := cmd.Run(); err != nil {
```

Comment:
I find it kinda awkward that commands store a reference to the bot when the bot is always in control of commands. It's an unneeded reference. Isn't it better to not convolute the command structures with the extra reference when the executing bot should be in control anyways and pass itself to `Run`? Binding bot to the command makes it less flexible for command reuse no and future adaptability, no?

Like `cmd.RunWithBot(b)`.

Reply:
YES. OF COURSE. I will do this.